ルートモーションの変換方法の修正

ke456-png opened the pull request:
ルートモーションを変換する際のアニメーショントラックに含まれるオフセットを除去する処理を変更します。


#  背景

HipsのpositionアニメーションにはスケルトンのHipsの初期位置がオフセットとして含まれておりそれを除去する必要がありました。
しかし、BVHのスケルトンを接地させる処理の影響でHipsの初期位置が正しく参照出来ていませんでした。


接地処理を廃止し、Hipsのpositionトラックに含まれるオフセットをスケルトンのHipsの初期位置によって削除するように変更しました。

ke456-png committed:
ルートモーションの変換方法の修正

BVHのスケルトンを接地させる処理の影響でルートアニメーションのオフセットを正しく除去出来ていなかった。

接地処理を削除し、Hipsのpositionトラックに含まれるオフセットをスケルトンのHipsの初期位置によって削除する処理を追加する。

0b5vr commented:
CMUモーションが正常に変換できないことを確認しました。
https://github.com/mrdoob/three.js/blob/dev/examples/models/bvh/pirouette.bvh

Model内で `this.vrm?.humanoid.getNormalizedBone('hips')?.node.position` を出力してみたところ、hipsのpositionがInfinityに吹っ飛んでいってそうです。

ke456-png committed:
バウンディングボックスの底辺が接地より下の場合、ルート位置を修正する

BVHはHips等にオフセットを持ちアニメーションを相対座標で持っているものと、
オフセットが0でアニメーションは絶対座標で記録されている場合があります。

そのためroot座標の処理は必要でした。

靴底の高さの分だけつま先ボーンが浮いていたりするため、バウンディングボックスの底辺が接地面とは限りません。
よって、バウンディングボックスの底辺が接地面より上の場合はそのオフセットを尊重し、
バウンディングボックスの底辺が接地より下の場合、接地面と同じ高さになるようにルート位置を修正します。

ke456-png commented:
BVHはHips等にオフセットを持ちアニメーションを相対座標で持っているものと、オフセットが0でアニメーションは絶対座標で記録されている場合があります。

VRMA変換時にスケルトンのHipsの高さを接地面を基準とするためにはroot座標の処理は必要でした。

ただし、靴底の高さの分だけつま先ボーンが浮いていたりするため、バウンディングボックスの底辺が接地面とは限りません。よって、バウンディングボックスの底辺が接地面より上の場合はそのオフセットを尊重し、バウンディングボックスの底辺が接地より下の場合、接地面と同じ高さになるようにルート位置を修正します。

0b5vr commented:
CMUモーション正常に変換できていそうです。

> よって、バウンディングボックスの底辺が接地面より上の場合はそのオフセットを尊重し、バウンディングボックスの底辺が接地より下の場合、接地面と同じ高さになるようにルート位置を修正します。

なるほど。モーションによって、キャラクターが少し宙に浮いてしまったり・逆に埋まったりしてしまう問題は、この差異によって発生している感じでしょうか。
靴底問題悩ましいですね……
